lib/Data/Decimate.pm

```diff
@@ -0,0 +1,235 @@
+
+use 5.010;
+use Moose;
+
+use Date::Utility;
+
+use MooseX::Types::Moose qw(Int Num Str);
+use MooseX::Types -declare => [qw(
+        interval
+        )];
+
+use Moose::Util::TypeConstraints;
+use Time::Duration::Concise;
+
+subtype 'interval', as 'Time::Duration::Concise';
+coerce 'interval', from 'Str', via { Time::Duration::Concise->new(interval => $_) };
+
+=head1 NAME
+
+Data::Decimate - A module that allows to decimate a data feed. 
+
+=head1 VERSION
+
+Version 0.01
+
+=head1 SYNOPSIS
+
+  use Data::Decimate;
+
+  my $data_decimate = Data::Decimate->new;
+
+  my @data_feed = [
+        {epoch  => time,
+        ...},
+        {epoch  => time+1,
+        ...},
+        {epoch  => time+2,
+        ...},
+        ...
+  ];
+
+  my $output = $data_decimate->decimate(\@data_feed);
+
+
+=head1 DESCRIPTION
+
+A module that allows you to resample a data feed
+
+=cut
+
+our $VERSION = '0.01';
+
+=head1 ATTRIBUTES
+=cut
+
+=head2 sampling_frequency
+
+=cut
+
+has sampling_frequency => (
+    is      => 'ro',
+    isa     => 'interval',
+    default => '15s',
+    coerce  => 1,
+);
+
+=head1 SUBROUTINES/METHODS
+
+=head2 _check_missing_data
+
+=cut
+
+sub _check_missing_data {
+    my ($self, $args) = @_;
+
+    my $decimate_data = $args->{data};
+
+    my @sorted_data = sort { $a <=> $b } keys %$decimate_data;
+    my $first_key   = $sorted_data[0];
+    my $last_key    = $sorted_data[-1];
+
+    if (scalar(@sorted_data)) {
+        for (my $i = $first_key; $i <= $last_key; $i = $i + $self->sampling_frequency->seconds) {
+            my $data = $decimate_data->{$i};
+
+            if (not $data) {
+                my $data     = $decimate_data->{$i - $self->sampling_frequency->seconds};
+                my %to_store = %$data;
+                $to_store{resample_epoch} = $i;
+                $to_store{count}          = 0;
+                $decimate_data->{$i}      = \%to_store;
+            }
+        }
+    }
+
+    return $decimate_data;
+}
+
+=head2 decimate
+
+Decimate a given data based on sampling frequency.
+
+=cut
+
+sub decimate {
+    my ($self, $args) = @_;
+
+    my $data = $args->{data};
+
+    my $ai = $self->sampling_frequency->seconds;    #default 15sec
+
+    my $counter             = 0;
+    my $prev_decimate_epoch = 0;
+    my %decimate_data;
+
+    if ($data) {
+        %decimate_data = map {
+            my $decimate_epoch = ($_->{epoch} % $ai) == 0 ? $_->{epoch} : $_->{epoch} - ($_->{epoch} % $ai) + $ai;
+            $counter = ($decimate_epoch == $prev_decimate_epoch) ? $counter + 1 : 1;
+            $_->{count}          = $counter;
+            $_->{decimate_epoch} = $decimate_epoch;
+            $prev_decimate_epoch = $decimate_epoch;
+            ($decimate_epoch) => $_
+        } @$data;
+    }
+
+    my $res = $self->_check_missing_data({
+        data => \%decimate_data,
+    });
+
+    my @sorted_data = sort { $a <=> $b } keys %$res;
+
+    my @vals = map { $res->{$_} } @sorted_data;
+    return \@vals;
+}
```

Comment:
It is not clear (from documentation ) what do you return? Just values or the initial samples? The example in the synopsis should clarify that.